Add breadcrumbs for categories with identical names

zuk3975 opened the pull request:
<!-----------------------------------------------------------------------------
Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows below with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop.com/8/contribute/contribution-guidelines/#pull-requests
------------------------------------------------------------------------------>

| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | When multiple categories with same name exists in shop then categories component in product page shows shortened breadcrumb tags instead of name. Breadcrumb should only contain as many parents until its unique. So for example if there are categories `Home > Accessories` and `Home > Accessories > Accessories` then breadcrumb should be `Accessories > Accessories`, but if there is additional category like `Home > Accessories > Accessories > Accessories`, then breadcrumb should be `Accessories > Accessories > Accessories`. If showing whole breadcrumb still doesn't provide uniqueness, then category id is appended to the end (e.g. `Home > Accessories (#4)`) Same principle applies to default category selection. This is to allow merchant to identify the category (its possible there will be categories with same names from different parents, so that should help). Also additional little issue fixed - when new default category is selected, x" icon to delete previous default category appears (for some reason native javascript "change" event didin't work before for defaultCategoryInput so i just switched to using jquery and it works)  ![Screenshot from 2022-05-18 15-44-50](https://user-images.githubusercontent.com/31609858/169075386-1496726b-bedf-450f-9902-c3ea7ae1cbb6.png)
| Type?             | improvement
| Category?         | BO
| BC breaks?        | see bellow
| Deprecations?     | no
| Fixed ticket?     | #28995
| Related PRs       | 
| How to test?      | :warning: Contains lots of js, so don't forget to recompile the assets for BO. Create multiple categories with identical name in same shop. Open experimental product page, description tab. In tags part categories that has same name should be shown with a optimized breadcrumb instead of name (Please read description for more info)
| Possible impacts? | 

**BC breaks**
* PrestaShop\PrestaShop\Core\Domain\Product\QueryResult\CategoryInformation array $localizedNames replaced with string $name. Added additional constructor argument string $displayName.
* PrestaShopBundle\Controller\Admin\Sell\Catalog\CategoryController::getCategoriesTreeAction() returns additional string "displayName" in categories array.

### Specs
https://github.com/PrestaShop/prestashop-specs/blob/master/content/1.7/back-office/catalog/products/add-edit-v2.md#categories

**Categories**

- **Associated categories**:
The component displays all the associated categories as a list of tags. If categories have the same name, the tags display the parent’s path association until the parents are different. If 2 categories have the same complete tree structure then it displays its category IDs and all its parents categories.

EX:
Product: Comics
Categories 1: US comics > Best-sellers > Thor > Marvel
Categories 2 : Home > Classical > Thor > Marvel
If the product is associated with categories 1 and 2
then the tags will display:
Tag1: Best-sellers > Thor > Marvel
Tag2: Classical > Thor > Marvel

Deleting a tag, removes the tag of the list and deletes the association between the category and the product at the save.

- **Add categories:**
Below the tags’ lists is displayed the “Add categories” button. The button opens a modal with the search component and the category tree.

The user can search by category and the parent’s categories. The search result list displays the breadcrumb of the category.  
The list of categories associated is also displayed below the search component. 

If the product is not associated with any category except the ‘Home’ branch, then ‘Home’’s child categories are displayed, just the first level of the tree is displayed/opened.

If the product is already associated with categories, then the categories branches are opened until the categories associated are displayed.

- **Main category** - Drop-list - It displays the full path of the associated categories. By default, at the creation of the product, the ‘Home’ category is selected as the main category. If categories have the same name, the tags display the parent’s path association until the parents are different. If 2 categories have the same complete tree structure then it displays its category IDs and all its parents categories.


The main category defines the product URL and the breadcrumb of the front office.

Below the drop-list is display the breadcrumb of the product. Updating the main category updates the breadcrumb displayed below the list and the breadcrumb of the front office.
<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/28141)
<!-- Reviewable:end -->

zuk3975 committed:
get breadcrumbs from api

zuk3975 committed:
return breadcrumbs in product for editing

zuk3975 committed:
show 2 breadcrumb parts if names matches on pageload

zuk3975 committed:
fix breadcrumb logic in categoryTree query

zuk3975 committed:
use category repository to get duplicate categories instead of php and use displayName var. All working needs polishing

jolelievre commented on admin-dev/themes/new-theme/js/pages/product/components/categories/types/index.d.ts:
```diff
@@ -12,6 +13,8 @@ type TypeaheadCategory = {
   name: string,
+  displayName: string,
   active: boolean,
+  breadcrumb: string,
```
This one is not used it can be cleaned

jolelievre commented on src/Adapter/Category/QueryHandler/GetCategoriesTreeHandler.php:
```diff
@@ -58,38 +73,61 @@ public function handle(GetCategoriesTree $query): array
         $langId = $query->getLanguageId() ? $query->getLanguageId()->getValue() : (int) $this->contextLangId;
         $nestedCategories = Category::getNestedCategories(null, $langId, false);
+        //@todo; hardcoded shop id. Should I add shop constraint to query, or take context shop id?
+        $this->duplicateCategoryNames = $this->categoryRepository->getDuplicateNames(new ShopId(1), $query->getLanguageId());
```
Yep it's better to anticipate this now instead of finding a bug later that we wouldn't understand at first

jolelievre commented on src/Adapter/Category/QueryHandler/GetCategoriesTreeHandler.php:
```diff
@@ -58,38 +73,61 @@ public function handle(GetCategoriesTree $query): array
-                $categoryChildren = $this->buildCategoriesTree($category['children'], $langId);
+                $parentCategories = $parents;
+                $parentCategories[] = $category;
+                $categoryChildren = $this->buildCategoriesTree($category['children'], $langId, $parentCategories);
```
```suggestion
                $categoryChildren = $this->buildCategoriesTree($category['children'], $langId, array_merge($parents, [$category]));
```

jolelievre commented on src/Adapter/Category/QueryHandler/GetCategoriesTreeHandler.php:
```diff
@@ -58,38 +73,61 @@ public function handle(GetCategoriesTree $query): array
             $categoryActive = (bool) $category['active'];
             $categoryChildren = [];
 
+            $breadcrumbs = [];
```
Not sure this `$breadcrumbs` variable is required, you can build the displayName almost the way by relying on `$parents` only

jolelievre commented on src/Adapter/Category/QueryHandler/GetCategoriesTreeHandler.php:
```diff
@@ -58,38 +73,61 @@ public function handle(GetCategoriesTree $query): array
+            return $name;
+        }
+
+        return implode(' > ', array_slice($breadcrumbs, -2, 2));
```
Does it work if the category doesn't have enough parents?

jolelievre replied:
You also need to put a very clear comment about why see use only two levels and with an example of the expected result here

jolelievre commented on src/Adapter/Product/QueryHandler/GetProductForEditingHandler.php:
```diff
@@ -262,6 +271,9 @@ private function getBasicInformation(Product $product): ProductBasicInformation
     {
+        $languageId = new LanguageId($this->contextLangId);
+        //@todo: hardcoded shop Id, do I inject context shop or require shop constraint?
+        $duplicateNames = $this->categoryRepository->getDuplicateNames(new ShopId(1), $languageId);
```
You should rely on `$query->getShopConstraint()` however it can be an `AllShops` constraint, in which case you need to use the product's default shop See in `ProductMultiShopRepository::getByShopConstraint` how it's handled

jolelievre commented on src/Adapter/Product/QueryHandler/GetProductForEditingHandler.php:
```diff
@@ -274,12 +286,38 @@ private function getCategoriesInformation(Product $product): CategoriesInformati
+
+        $breadcrumbs = $this->categoryRepository->getBreadcrumbParts($categoryId, $languageId);
+
+        return implode(' > ', array_slice($breadcrumbs, -2, 2));
```
Same here as in the tree handler this needs to be documented with much details because there's no way someone reaches this code and understands the purpose out of the blue

jolelievre commented on src/Core/Domain/Category/QueryResult/CategoryForTree.php:
```diff
@@ -98,4 +123,12 @@ public function getChildren(): array
+    /**
+     * @return array<int, string[]>
+     */
+    public function getLocalizedBreadcrumbs(): array
```
I believe this is not used either this part should be cleaned

jolelievre commented on src/Core/Domain/Product/QueryResult/CategoryInformation.php:
```diff
@@ -37,21 +37,31 @@ class CategoryInformation
     /**
      * @var array<int, string>
+     * @todo: this doesn't need to be localized because we are not updating category info in product page
+     *        (product name, description are localized, because all languages are needed for update)
```
You can refacto this one with a single string If the query doesn't handle a LanguegId and the expected result is only one language there is no reason to keep some mixed data (some localized some not) So it's better to clearly state the contract of this query
BC break is ok here

jolelievre commented on src/PrestaShopBundle/Controller/Admin/Sell/Catalog/CategoryController.php:
```diff
@@ -801,12 +801,15 @@ private function formatCategoriesTreeForPresentation(array $categoriesTree, int
                 'name' => $names[$langId] ?? reset($names),
+                'displayName' => $categoryForTree->getDisplayName(),
                 'children' => $children,
+                'breadcrumbs' => $breadcrumbs[$langId] ?? reset($breadcrumbs),
```
This can be removed as well

jolelievre commented on src/PrestaShopBundle/Form/Admin/Type/TextPreviewType.php:
```diff
@@ -48,6 +48,7 @@ public function buildView(FormView $view, FormInterface $form, array $options)
         parent::buildView($view, $form, $options);
         $view->vars['type'] = 'hidden';
         $view->vars['preview_class'] = $options['preview_class'];
+        $view->vars['preview_value'] = $options['custom_preview_value'] ?? $form->getData();
```
I don't think you need this either

jolelievre commented on src/PrestaShopBundle/Resources/views/Admin/TwigTemplateForm/prestashop_ui_kit_base.html.twig:
```diff
@@ -1251,7 +1251,7 @@
 
     {# This just outputs the value #}
-    {{ form.vars.value|e('html') }}
+    {{ form.vars.preview_value|e('html') }}
```
Not used apparently so it should be cleaned

jolelievre reviewed (commented):
Hi @zuk3975 

overall it seems fine but I believe there is a lot of unused code, maybe your first attempt on another solution?

zuk3975 committed:
single lang name in categoryInformation

zuk3975 committed:
fix ProductFormDataProviderTest

zuk3975 committed:
remove localized names from CategoryTree

jolelievre commented on src/Adapter/Category/QueryHandler/GetCategoriesTreeHandler.php:
```diff
@@ -77,19 +103,53 @@ private function buildCategoriesTree(array $categories, int $langId): array
     }
+
+    /**
+     * If there are multiple categories with identical names, we want to be able to tell them apart,
```
Great description! We covered our asses 😅

jolelievre commented on src/Adapter/Product/QueryHandler/GetProductForEditingHandler.php:
```diff
@@ -274,12 +288,54 @@ private function getCategoriesInformation(Product $product): CategoriesInformati
+     *
+     * @return string
+     */
+    private function buildDisplayName(
```
I think this technical feature could be handled in a separate class because you do almost the same thing in the tree handler. I imagine a service `CategoryDisplayNameBuilder` which would internally fetch the `duplicateNames` by using `Category::getDuplicateNames` internally. And a single public method:

```php
public function buidDisplayName(string $categoryName, bool $refreshDuplicates = false): string
```

By default, the duplicated names is fetched once and stored in a class field, but the `$refreshDuplicates` can force fetching it. I think about this for the test issues, because we need some kind of cache that can be cleared easily. Maybe you should use a cache adapter or the legacy `Cache` class but I'm less fan of the legacy approach.

jolelievre reviewed (commented):
Almost all good, I just thought of an improvement for combination building Actually maybe this feature could be added to the already existing `CombinationNameBuilder` service The `CombinationNameBuilderInterface` is one so we can update it as well

Progi1984 commented on src/Adapter/Category/QueryHandler/GetCategoriesTreeHandler.php:
```diff
@@ -42,54 +47,102 @@ final class GetCategoriesTreeHandler implements GetCategoriesTreeHandlerInterfac
      * @return CategoryForTree[]
      */
-    private function buildCategoriesTree(array $categories, int $langId): array
+    private function buildCategoriesTree(array $categories, ShopId $shopId, LanguageId $langId, $parents = []): array
```
```suggestion
    private function buildCategoriesTree(array $categories, ShopId $shopId, LanguageId $langId, array $parents = []): array
```

Progi1984 commented on src/Adapter/Category/Repository/CategoryRepository.php:
```diff
@@ -140,6 +141,40 @@ public function assertCategoryExists(CategoryId $categoryId): void
         }
     }
 
+    public function getDuplicateNames(ShopId $shopId, LanguageId $languageId): array
```
Please add PhpDoc

jolelievre replied:
Indeed it might be useful to avoid reverse engineering to document and explain why we need this, since this feature is not really intuitive at first

Progi1984 commented on src/Core/Category/NameBuilder/CategoryInformation.php:
```diff
@@ -0,0 +1,67 @@
+     */
+    private $name;
+
+    public function __construct(
```
Please add phpdoc

Progi1984 commented on src/Core/Domain/Category/QueryResult/CategoryForTree.php:
```diff
@@ -52,18 +57,21 @@ class CategoryForTree
         int $categoryId,
         bool $active,
-        array $localizedNames,
+        string $name,
+        string $displayName,
```
Please define the BC Break

jolelievre replied:
We don't mark BC break on product page related code anymore because it was all considered experimental

Progi1984 replied:
Sorry, but the class is not defined in `1.7.8.x` as `@internal` : https://github.com/PrestaShop/PrestaShop/blob/1.7.8.x/src/Core/Domain/Category/QueryResult/CategoryForTree.php

Please define BC Breaks.

Progi1984 reviewed (dismissed):
@zuk3975 Thank you for your contribution. Could you add behat tests for checking breadcrumbs for categories with identical names ?

jolelievre commented on src/Adapter/Category/QueryHandler/GetCategoriesTreeHandler.php:
```diff
@@ -42,54 +47,102 @@ final class GetCategoriesTreeHandler implements GetCategoriesTreeHandlerInterfac
     public function __construct(
-        string $contextLangId
+        string $contextLangId,
+        int $contextShopId,
```
I believe I said otherwise earlier but on second thought I think this should be in the query object via a shop constraint.

It doesn't need to be handled perfectly because it's not part of the full multishop PR for this feature, but the idea is to anticipate the structure of this `GetCategoriesTree` query. Since multishop is on pause we are not sure to finish everything for the 8.0

But at least if this query already has the proper structure it limits the breaking changes that we will need to add in the next version. Thus the require work here is not that huge:
- add shop constraint in the query object
- in the CategoryController build this query based on the context shop ID (use default shop as fallback if it's not set, but since the product page is only accessible in single shop anyway it shouldn't matter)

Basically, the idea is to move the injection of shop ID from here to the controller via the query object. I know you don't like to increase the scope of PR but this one change is not that big and you already handled the shop ID partially, so is it ok for you? I won't block the PR for this anyway

jolelievre commented on src/Adapter/Category/QueryHandler/GetCategoriesTreeHandler.php:
```diff
@@ -42,54 +47,102 @@ final class GetCategoriesTreeHandler implements GetCategoriesTreeHandlerInterfac
-        $langId = $query->getLanguageId() ? $query->getLanguageId()->getValue() : (int) $this->contextLangId;
-        $nestedCategories = Category::getNestedCategories(null, $langId, false);
+        $shopId = new ShopId((int) $this->contextShopId);
+        $langId = $query->getLanguageId() ?? new LanguageId((int) $this->contextLangId);
```
Seems like the `$languageId` of `GetCategoriesTree` is optional, so the fallback is handled in this handler AND in the controller that uses the query. It seems a bit redundant so I think we should force the parameter and not allow it to be null anymore.

I guess this was done to avoid a breaking change initially but we shouldn't be bound by this constraint for a major version and experimental code. So we can clean this and make the handler lighter.

jolelievre commented on src/Core/Category/NameBuilder/CategoryDisplayNameBuilder.php:
```diff
@@ -0,0 +1,130 @@
+    private function buildCacheKey(ShopId $shopId, LanguageId $langId): string
+    {
+        return sprintf(
+            'duplicateCategoryNames_shop_%s_lang_%s',
```
```suggestion
            'Category::duplicateCategoryNames_shop_%s_lang_%s',
```

jolelievre replied:
This way this cache can be flushed along with the rest of category cache doing `Cache::clean('Category::*')` which should be added in `Category::resetStaticCache`

jolelievre commented on src/Adapter/Category/QueryHandler/GetCategoriesTreeHandler.php:
```diff
@@ -42,54 +47,102 @@ final class GetCategoriesTreeHandler implements GetCategoriesTreeHandlerInterfac
-        $nestedCategories = Category::getNestedCategories(null, $langId, false);
+        $shopId = new ShopId((int) $this->contextShopId);
+        $langId = $query->getLanguageId() ?? new LanguageId((int) $this->contextLangId);
+        $nestedCategories = Category::getNestedCategories(null, $langId->getValue(), false);
```
I believe this call to static method `getNestedCategories` needs to be wrapped in a `try/finally` block with the context state manager in order to get the appropriate content and names (since the shop condition is handled internally with legacy conditions code)

jolelievre commented on src/Core/Category/NameBuilder/CategoryInformation.php:
```diff
@@ -0,0 +1,67 @@
+
+use PrestaShop\PrestaShop\Core\Domain\Category\ValueObject\CategoryId;
+
+class CategoryInformation
```
This should be renamed to avoid confusing it with `PrestaShop\PrestaShop\Core\Domain\Product\QueryResult\CategoryInformation` even if they have different namespaces

jolelievre replied:
Actually, I'm not sure this class is needed at all, we are forced to create this at each call of the builder Always just in time in the method call itself, the object is not stored and reused in any ways

So it might be better to remove this class and simply use to parameters in the `build(CategoryId $id, string $name, ...)` method

jolelievre commented on src/Core/Category/NameBuilder/CategoryDisplayNameBuilder.php:
```diff
@@ -0,0 +1,130 @@
+use PrestaShop\PrestaShop\Core\Domain\Language\ValueObject\LanguageId;
+use PrestaShop\PrestaShop\Core\Domain\Shop\ValueObject\ShopId;
+
+class CategoryDisplayNameBuilder
```
```suggestion
/**
 * This class is used to build category display name, mostly to avoid confusing categories which have the same names.
 */
class CategoryDisplayNameBuilder
```

jolelievre replied:
Just a small description for the class itself, it's complimentary with the function description which gives more details.

jolelievre commented on src/Adapter/Form/ChoiceProvider/ProductDefaultCategoryChoiceProvider.php:
```diff
@@ -45,16 +49,40 @@ class ProductDefaultCategoryChoiceProvider implements ConfigurableFormChoiceProv
         $this->defaultCategoryId = $homeCategoryId;
         $this->categoryRepository = $categoryRepository;
+        $this->categoryDisplayNameBuilder = $categoryDisplayNameBuilder;
+        $this->shopId = $shopId;
+        $this->languageId = $languageId;
```
```suggestion
        $this->shopId = new ShopId($shopId);
        $this->languageId = new LanguageId($languageId);
```

jolelievre replied:
This will avoid creating new objects at each call of `getDisplayName`

jolelievre commented on src/Adapter/Product/QueryHandler/GetProductForEditingHandler.php:
```diff
@@ -185,6 +202,8 @@ public function __construct(
         $this->productImageUrlFactory = $productImageUrlFactory;
         $this->specificPriceRepository = $specificPriceRepository;
         $this->configuration = $configuration;
+        $this->contextLangId = $contextLangId;
```
On second thought, I'm sorry I have a lot of second thoughts in this review 😅, it would be interesting to move the `languageId` inside the query itself. It means the contextLangId would be injected in the `ProductFormDataProvider` to create the query object (and actually it already has it)

And it would allow behat testing this query with different languages. SO I would suggest adding a `$displayLanguageId` parameter to the query

And it's usually better when we can avoid injecting contextual values in the handler (when it's possible easily)

jolelievre commented on tests/Integration/Behaviour/Features/Context/Domain/CategoryFeatureContext.php:
```diff
@@ -115,6 +116,8 @@ public function assertRootCategoriesTree(TableNode $tableNode, string $langIso):
      */
     public function assertCategoriesTree(TableNode $tableNode, string $langIso, string $parentReference): void
     {
+        Cache::clean('Category::getNestedCategories_*');
```
This should be moved into `Category::resetStaticCache` (needs to be created) as it can be useful in other test use cases

jolelievre reviewed (commented):
Thanks @zuk3975 

A few architectural suggestions (sorry I changed my mind on a few decisions we already talked about 😅) but it's looking good!

jolelievre commented:
Also @Progi1984 which behat tests do you ask? Because it seems like tests for product data and category tree have all been updated already

zuk3975 committed:
fix ProductDefaultCategoryChoiceProvider

 require shopId in tree query

 wrap nestedCategories call in try/finally

zuk3975 commented:
> I think I've resolved everything except [#28141 (comment)](https://github.com/PrestaShop/PrestaShop/pull/28141#discussion_r854037053) (this requires adding additional behat scenarios for multiple shops). So PR is not ready until that is done.

ping @jolelievre okay, actually im afraid that adding these behats will raise the scope too much, as I would need to introduce a way to create categories for different shops, so if thats okay, im not adding additional tests to support categoryTree multishop behavior in this pr?

MeKeyCool commented on src/Core/Category/NameBuilder/CategoryDisplayNameBuilder.php:
```diff
@@ -0,0 +1,136 @@
+            );
+        }
+
+        return implode(' > ', array_slice($breadcrumbParts, -2, 2));
```
Isn't this code redundant with `CategoryRepository::getBreadcrumb(...)` ? Why not adding number of parents in `getBreadcrumb` arguments ?

jolelievre replied:
@MeKeyCool I found this suggestion interesting at first but then I thought if we add a parameter to fetch only the last elements, then we would need to have another to get the first elements Then it becomes a function with two optional parameters and it becomes messy So I think the current implementation is enough

At worst we could have alternative methods like `getBeginningBreadcrumb` / `getEndingBreadcrumb` where the parameter would not be optional, but it becomes slightly overengineered

jolelievre commented on src/Adapter/Category/QueryHandler/GetCategoriesTreeHandler.php:
```diff
@@ -28,68 +28,129 @@
-        $nestedCategories = Category::getNestedCategories(null, $langId, false);
+        $langId = $query->getLanguageId();
+        $this->contextStateManager->saveCurrentContext();
+        $this->contextStateManager->setShop($this->shopRepository->get($query->getShopId()));
```
```suggestion
        $this->contextStateManager
            ->saveCurrentContext()
            ->setShop($this->shopRepository->get($query->getShopId()))
        ;
```

jolelievre replied:
Just a suggestion, not blocking
It's personal, I like fluent setters ^^

jolelievre commented on src/Core/Category/NameBuilder/CategoryDisplayNameBuilder.php:
```diff
@@ -0,0 +1,136 @@
+     *
+     * @return string
+     */
+    public function build(
+        CategoryId $categoryId,
+        string $categoryName,
+        ShopId $shopId,
+        LanguageId $languageId,
+        ?array $breadcrumbParts = null,
+        bool $useCache = true
+    ): string {
```
```suggestion
    public function build(
        CategoryId $categoryId,
        string $categoryName,
        ShopId $shopId,
        LanguageId $languageId,
        bool $useCache = true
    ): string {
```

jolelievre replied:
I'm still not fully convinced by this extra optional `breadcrumb` parameter (even if we discussed this together already ^^) Although I totally agree with its purpose to optimize/avoid the breadcrumb fetching when it's available

So to fit both needs and keep the prototype cleaner I suggest having two methods `build` and `buildWithBreadcrumbs` that each handle a different use case

zuk3975 replied:
yes I agree with you, but I don;t have any better offers :sweat:

jolelievre commented on src/Core/Category/NameBuilder/CategoryDisplayNameBuilder.php:
```diff
@@ -0,0 +1,136 @@
+
+        $cacheKey = $this->buildCacheKey($shopId, $languageId);
+
+        if (Cache::isStored($this->buildCacheKey($shopId, $languageId))) {
```
This is optional for this PR, but on second thought it would be better if we could rely on `Symfony\Component\Cache\Adapter\AdapterInterface` here instead of the legacy Cache

But it's not mandatory in this PR, so could you at least add a `@todo` so that we refacto this later?

zuk3975 replied:
reminder issue https://github.com/PrestaShop/PrestaShop/issues/28427

jolelievre commented on src/Core/Domain/Product/Query/GetProductForEditing.php:
```diff
@@ -45,17 +46,23 @@ class GetProductForEditing
+    /**
      * @param int $productId
      * @param ShopConstraint $shopConstraint
+     * @param int $languageId
```
```suggestion
     * @param int $displayLanguageId
```

jolelievre commented on src/Core/Form/IdentifiableObject/DataProvider/ProductFormDataProvider.php:
```diff
@@ -107,7 +107,9 @@ public function getData($id): array
         /** @var ProductForEditing $productForEditing */
-        $productForEditing = $this->queryBus->handle(new GetProductForEditing($productId, $shopConstraint));
+        $productForEditing = $this->queryBus->handle(
+            new GetProductForEditing($productId, $shopConstraint, $this->contextLangId)
```
This is great!!

jolelievre commented on tests/Integration/Behaviour/Features/Context/Domain/AbstractDomainFeatureContext.php:
```diff
@@ -293,6 +294,14 @@ protected function localizeByCell(string $localizedValue): array
+    /**
+     * @return int
+     */
+    protected function getContextShopId(): int
```
We shouldn't integrate anything fetched from the Context in our behat tests **ever**

The reason is that we are currently "struggling" to change the command format to force them and include some ShopConstraint exactly because we don't want to rely on the Context anymore So if we get rid of it somewhere it's not to introduce it again somewhere else

The command are now built to have a full knowledge of their own context, so their tests should as well. So this should should be removed.

When you're not sure which shopId to use in your behat tests that are oriented as single shop you can usually use the default shop.

You can use `AbstractProductFeatureContext::getDefaultShopId()` in those case (maybe one day we'll have to move it to a lower layer like `AbstractDomainFeatureContext` but for now it's mainly useful for product commands which are the only ones handling shop constraints (for now)

jolelievre replied:
Ah I just noticed that you need this in `CategoryFeatureContext` which doesnextend the product class, so I guess now is a good time to move `getDefaultShopId` to `AbstractDomainFeatureContext` ^^

jolelievre commented on tests/Integration/Behaviour/Features/Context/Domain/CategoryFeatureContext.php:
```diff
@@ -99,11 +99,11 @@ public function __construct()
     {
         $langId = Language::getIdByIso($langIso);
-        $categoriesTree = $this->getQueryBus()->handle(new GetCategoriesTree($langId));
+        $categoriesTree = $this->getQueryBus()->handle(new GetCategoriesTree($langId, $this->getContextShopId()));
```
```suggestion
        $categoriesTree = $this->getQueryBus()->handle(new GetCategoriesTree($langId, $this-> getDefaultShopId()));
```

jolelievre commented on tests/Integration/Behaviour/Features/Context/Domain/CategoryFeatureContext.php:
```diff
@@ -115,15 +115,16 @@ public function assertRootCategoriesTree(TableNode $tableNode, string $langIso):
+        Category::resetStaticCache();
         $langId = Language::getIdByIso($langIso);
-        $categoriesTree = $this->getQueryBus()->handle(new GetCategoriesTree($langId));
+        $categoriesTree = $this->getQueryBus()->handle(new GetCategoriesTree($langId, $this->getContextShopId()));
```
```suggestion
        $categoriesTree = $this->getQueryBus()->handle(new GetCategoriesTree($langId, $this-> getDefaultShopId()));
```

jolelievre reviewed (changes requested):
Thanks @zuk3975 

Just a few suggestions for refacto but it's looking good! Despite the feature itself, this PR brings a lot of nice technical improvements it's great!

Note: especially the test part needs to be fixed. We don't want the `Context` singleton to create a mess in the behat tests

zuk3975 committed:
extract display name builder to 2methods removing breadcrumbs from one

jolelievre commented on src/Core/Category/NameBuilder/CategoryDisplayNameBuilder.php:
```diff
@@ -0,0 +1,183 @@
+     *
+     * @return string
+     */
+    private function buildDisplayName(
+        CategoryId $categoryId,
+        string $categoryName,
+        ShopId $shopId,
+        LanguageId $languageId,
+        ?array $breadcrumbParts = null,
+        bool $useCache = true
+    ): string {
```
```suggestion
    private function buildDisplayName(
        string $categoryName,
        ShopId $shopId,
        LanguageId $languageId,
        CategoryId $categoryId = null,
        ?array $breadcrumbParts = null,
        bool $useCache = true
    ): string {
```

jolelievre replied:
`$categoryId` and `$breadcrumbParts` are both optional since they both correspond to two different use cases:
- `build` : no breadcrumb provided, so we need the category ID to get it via the repository
- `buildWithBreadcrumbs`: breadcrumb is provided so no need for the category ID

jolelievre commented on src/Core/Category/NameBuilder/CategoryDisplayNameBuilder.php:
```diff
@@ -0,0 +1,183 @@
+     * @return string
+     */
+    public function buildWithBreadcrumbs(
+        CategoryId $categoryId,
```
```suggestion
```

jolelievre replied:
Seems like `$categoryId` is only used by `CategoryRepository::getBreadcrumbParts` so it's not useful for this method which already has breadcrumb parts provided, no need to keep it in this public method

jolelievre commented on src/Core/Category/NameBuilder/CategoryDisplayNameBuilder.php:
```diff
@@ -0,0 +1,183 @@
+     * @return string
+     */
+    public function build(
+        CategoryId $categoryId,
+        string $categoryName,
+        ShopId $shopId,
+        LanguageId $languageId,
+        bool $useCache = true
```
```suggestion
        string $categoryName,
        ShopId $shopId,
        LanguageId $languageId,
        CategoryId $categoryId,
        bool $useCache = true
```

jolelievre replied:
This one is just a suggestion, but since `$categoryId` is not useful in `buildWithBreadcrumbs` (see the comment below why) I though it might be interesting that the common parameters of the two methods are in the same order and the different parameter is placed last.

That said my suggestion also works (and maybe is more suited) if we place the different parameter first, to emphasize on the difference.

But in both case the idea is to have similar ways of using both methods. It's just an idea like this though, the specific order of parameters is not mandatory (however only keep useful parameters is important).

jolelievre commented on src/Core/Domain/Product/Query/GetProductForEditing.php:
```diff
@@ -45,17 +46,23 @@ class GetProductForEditing
+    /**
      * @param int $productId
      * @param ShopConstraint $shopConstraint
+     * @param int $displayLanguageId
```
```suggestion
     * @param int $displayLanguageId Some data used in the form are only here for the user knowledge and are not editable so they only need to be provided in the language of the interface, defined by this parameter.
```

jolelievre reviewed (commented):
Thanks @zuk3975 

The description are very clear, even for me who knew the PR's history it gave me better viewing of the code ^^ 

Only a comment suggested, and one last thing to improve in the name builder (clean useless parameter) and it's all good for me.

zuk3975 committed:
clean up

jolelievre reviewed (approved):
Thanks @zuk3975

Amoifr commented on src/Adapter/Category/Repository/CategoryRepository.php:
```diff
@@ -140,6 +141,52 @@ public function assertCategoryExists(CategoryId $categoryId): void
+
+        $results = $qb->execute()->fetchAllAssociative();
+
+        $names = [];
```
may be using array_column will be shorter

Amoifr commented on src/Adapter/Category/Repository/CategoryRepository.php:
```diff
@@ -140,6 +141,52 @@ public function assertCategoryExists(CategoryId $categoryId): void
+     *
+     * @return string[]
+     */
+    public function getBreadcrumbParts(CategoryId $categoryId, LanguageId $languageId): array
```
Why adding a public method that only call a private. Why don't making this private public ?

jolelievre replied:
You're right I didn't notice but these three methods are kind of overkill A single public method should be enough

Progi1984 replied:
I agree. 

@zuk3975 Could you fix it ?

zuk3975 replied:
I've already fixed it (got rid of one redundant private method). Not sure why you don't see it (unless there is something else Im missing?)

Amoifr commented on src/Adapter/Category/Repository/CategoryRepository.php:
```diff
@@ -149,8 +196,19 @@ public function assertCategoryExists(CategoryId $categoryId): void
      */
     public function getBreadcrumb(CategoryId $categoryId, LanguageId $languageId, string $separator = ' > '): string
```
for user customisation, it will be nice to inject the separator ?

jolelievre replied:
I don't think this should be injected in this because it's a Repository It already allows specifying a custom optional separator in the method so this one is fine

However, the separator could be injected in `CategoryDisplayNameBuilder` so it can use it as a parameter when it uses the repository

Amoifr commented on src/Adapter/Category/Repository/CategoryRepository.php:
```diff
@@ -184,13 +242,14 @@ public function getBreadcrumb(CategoryId $categoryId, LanguageId $languageId, st
+
         $parentNames = [];
-        foreach ($result as $category) {
+        foreach ($results as $category) {
```
array_column ?

zuk3975 replied:
ah nice, didn't actually know about this method :smile: :+1:

Amoifr commented on src/Core/Category/NameBuilder/CategoryDisplayNameBuilder.php:
```diff
@@ -0,0 +1,168 @@
+     */
+    private function buildBreadcrumb(array $breadcrumbParts): string
+    {
+        return implode(' > ', array_slice($breadcrumbParts, -2, 2));
```
separator can be provided by a conf ?

Amoifr replied:
cf my previous comment for separator in this PR. If we decide to change one we might forget to chance also here.

jolelievre replied:
In this class it makes more sense to allow customizing/injecting the separator

@zuk3975 the idea here is to inject a string separator which is hard-coded in the yaml service like in this example:

https://github.com/PrestaShop/PrestaShop/blob/422dccaa05b1e232014bf7cf21b29814693abe20/src/PrestaShopBundle/Resources/config/services/core/product.yml#L13-L17

jolelievre reviewed (changes requested):
Sorry @zuk3975 I missed some stuff that was legitimately hlighlited by @Amoifr 

Just two things to fix

jolelievre reviewed (dismissed):
Thanks @zuk3975

Progi1984 commented on classes/Category.php:
```diff
@@ -1391,6 +1391,15 @@ public function getName($idLang = null)
+    /**
+     * {@inheritdoc}
+     */
+    public static function resetStaticCache()
```
```suggestion
    public static function resetStaticCache(): void
```

Progi1984 commented on src/Adapter/Category/Repository/CategoryRepository.php:
```diff
@@ -140,17 +141,51 @@ public function assertCategoryExists(CategoryId $categoryId): void
+    public function getDuplicateNames(ShopId $shopId, LanguageId $languageId): array
+    {
+        $qb = $this->connection->createQueryBuilder()
+            ->select('cl.name, COUNT(cl.name) AS nameCount')
+            ->from($this->dbPrefix . 'category_lang', 'cl')
+            ->where('id_lang = :langId')
+            ->andWhere('id_shop = :shopId')
+            ->having('nameCount > 1')
```
```suggestion
            ->select('cl.name')
            ->from($this->dbPrefix . 'category_lang', 'cl')
            ->where('id_lang = :langId')
            ->andWhere('id_shop = :shopId')
            ->having('COUNT(cl.name) > 1')
```

Progi1984 reviewed (dismissed):
Remaing some feedback from me & @Amoifr

khouloudbelguith reviewed (commented):
Hi @zuk3975,

I found some issues

**First issue**
1. Go to BO > Catalog > Categories and create a category `Home` 
![image](https://user-images.githubusercontent.com/16067358/168573695-a43dfeb1-b624-4b1c-93c1-c083f0f032a6.png)
2. Go to BO > Catalog > Products V2 > Add a product
3. In the Categories section, add the new category created
4. Save
5. See error 
![image](https://user-images.githubusercontent.com/16067358/168574703-b9c73533-c6cc-4621-95dc-237f7120b170.png)

https://user-images.githubusercontent.com/16067358/168574764-4e28e713-e75d-4d7c-bf53-dcd421fd6d9f.mp4

**_Without Pr_**
![image](https://user-images.githubusercontent.com/16067358/168575028-55ff3776-879b-4bb8-b3b9-a87d8a2c276e.png)

https://user-images.githubusercontent.com/16067358/168575179-7eb7e907-f32e-4331-8db4-92424dd7a600.mp4

**Second issue**
1. Go to BO > Catalog > Categories and create two categories with the same name and with different/similar parent `Home > Classic > Thor` and `Home > Best-sellers > Thor` 
2. Go to BO > Catalog > Products V2 > Add a product
3. In the Categories section, add the new categories created 
![image](https://user-images.githubusercontent.com/16067358/168575979-973996e1-2aa1-4693-9b31-a15ece057ae1.png)
4. Apply
5. In the Main category - Drop-list - It does not display the full path of the associated categories
![image](https://user-images.githubusercontent.com/16067358/168576250-feacabe0-f933-413a-9feb-8f22019f2c66.png)

https://user-images.githubusercontent.com/16067358/168576347-f7b8336c-bb4b-4869-bc25-6de062c5ba7a.mp4

**Third issue**

1. Go to BO > Catalog > Categories and create two categories with the same name and with similar parent `Home > test` and `Home > test` 
![image](https://user-images.githubusercontent.com/16067358/168575539-3e358048-9489-4b5f-8e44-a1b537060818.png)
2. Go to BO > Catalog > Products V2 > Add a product
3. In the Categories section, add the new categories created 
![image](https://user-images.githubusercontent.com/16067358/168575740-7bfa3d80-7ec7-47d0-a68b-ceb194a20023.png)
4. Save
5. In the Main category - Drop-list - Only one category is displayed
![image](https://user-images.githubusercontent.com/16067358/168577149-e6a2aaa4-afc8-40a6-a2d7-8cd768342194.png)

https://user-images.githubusercontent.com/16067358/168577188-703cfeba-66da-4cd5-bfb3-7c5adccdb36a.mp4

**Fourth issue**

1. Go to BO > Catalog > Categories and create a category with the name `Category1`
2. Edit this category and rename it `Category1updated`
3. Go to BO > Catalog > Products V2 > Add a product
3. In the Categories section, add the new category created 
4. Save
5. See error: the category list > the old name of the category is displayed 
![image](https://user-images.githubusercontent.com/16067358/168578540-46f5c4f3-e2a9-434f-9487-be4dbbaaf7a6.png)

https://user-images.githubusercontent.com/16067358/168578656-ed37f047-d663-455b-a80b-5d822af3c6a1.mp4

**_WithoutPr_**

https://user-images.githubusercontent.com/16067358/168578938-34af10f4-75a4-4edc-8076-b16a3a98d702.mp4

**Fifth issue**
1. Create a disabled category
2. Go to BO > Catalog > Products V2 > Add a product
3. Click on Add categories
4. In the Modal
5. The disabled category is displayed in italic font-style 
![image](https://user-images.githubusercontent.com/16067358/168580217-f587d48c-fc01-4e1e-a493-e9c166cab697.png)

Isn't the wanted behavior? ping @MatShir what do you think?

**Sixth issue**
1. Enable multistore & create a new shop `shop2`
2. Go to BO > Catalog > Categories page
3. Default shop is selected > create a new category `shop` available only in the default shop
4. Shop2 is selected > create a new category `shop` available only in the shop2
5. Shop2 is selected > Go to BO > Catalog > Products V2 > Add a product
6. Click on Add categories & add the created category
7. Save
8. In the Catagories list & the In the Main category - Drop-list - The category is shown with the full path of the associated categories

![image](https://user-images.githubusercontent.com/16067358/168581612-19d0a1ad-4b3e-4db5-a6ae-d12251f4d834.png)

Is'nt a normal behavior?
![image](https://user-images.githubusercontent.com/16067358/168581706-6b2b2f81-8f42-4951-a7b0-f1482180ea7b.png)
Ping @MatShir what do you think?

Thanks!

jolelievre reviewed (changes requested):
Oh I approved a bit too quickly because there was no midf
But since there is no modif and still some QA feedback I guess it was not fixed yet ^^

zuk3975 committed:
fix event listener for default category

zuk3975 committed:
fix js dynamic rendering of category default choices

zuk3975 committed:
Build breadcrumbs depending on how many parents duplicates

zuk3975 commented:
not ready yet. In progress working with bugs

zuk3975 commented:
done :heavy_check_mark:

zuk3975 commented:
done :heavy_check_mark:

zuk3975 committed:
fix default category cross update after selection

PululuK commented on src/Core/Category/NameBuilder/CategoryDisplayNameBuilder.php:
```diff
@@ -0,0 +1,265 @@
+            }
+        }
+
+        if ($duplicatedFound) {
+            // if breadcrumbs are still duplicated append category id
+            $breadcrumb = sprintf('%s (#%d)', $breadcrumb, $categoryId->getValue());
+        }
```
IMO the category ID should be present in all ctg names, duplicateds or not.

zuk3975 replied:
That is for @MatShir to decide. But i don't think its very useful (maybe for developers yes, but probably not merchants)

PululuK commented on src/Core/Category/NameBuilder/CategoryDisplayNameBuilder.php:
```diff
@@ -0,0 +1,265 @@
+            }
+        }
+
+        if ($duplicatedFound) {
+            // if breadcrumbs are still duplicated append category id
+            $breadcrumb = sprintf('%s (#%d)', $breadcrumb, $categoryId->getValue());
+        }
```
IMO the category ID should be present in all ctg names, duplicateds or not.

zuk3975 replied:
Wasn't specified this way, so I guess we try current way and when/if we get some feedback about this we can improve :v:

PululuK replied:
Hi @zuk3975 
Sorry it is a old comment, yep look goo to me, thanks !

Progi1984 commented on src/Adapter/Form/ChoiceProvider/ProductDefaultCategoryChoiceProvider.php:
```diff
@@ -45,16 +48,40 @@ class ProductDefaultCategoryChoiceProvider implements ConfigurableFormChoiceProv
+    /**
+     * @var CategoryDisplayNameBuilder
+     */
+    private $categoryDisplayNameBuilder;
+
+    /**
+     * @var ShopId
+     */
+    private $shopId;
+
+    /**
+     * @var LanguageId
+     */
+    private $languageId;
```
```suggestion
    protected $categoryDisplayNameBuilder;

    /**
     * @var ShopId
     */
    protected $shopId;

    /**
     * @var LanguageId
     */
    protected $languageId;
```

Progi1984 commented on src/Adapter/Form/ChoiceProvider/ProductDefaultCategoryChoiceProvider.php:
```diff
@@ -85,12 +115,30 @@ private function getChoicesForExistingProduct(int $productId): array
+     *
+     * @return string
+     */
+    private function getDisplayName(CategoryId $categoryId, string $categoryName): string
```
```suggestion
    protected function getDisplayName(CategoryId $categoryId, string $categoryName): string
```

Progi1984 commented on src/Adapter/Product/QueryHandler/GetProductForEditingHandler.php:
```diff
@@ -141,6 +143,11 @@ final class GetProductForEditingHandler implements GetProductForEditingHandlerIn
+    /**
+     * @var CategoryDisplayNameBuilder
+     */
+    private $categoryDisplayNameBuilder;
```
```suggestion
    protected $categoryDisplayNameBuilder;
```

Progi1984 commented on src/Adapter/Product/QueryHandler/GetProductForEditingHandler.php:
```diff
@@ -257,24 +267,34 @@ private function getBasicInformation(Product $product): ProductBasicInformation
      * @return CategoriesInformation
      */
-    private function getCategoriesInformation(Product $product): CategoriesInformation
+    private function getCategoriesInformation(Product $product, LanguageId $languageId): CategoriesInformation
```
```suggestion
    protected function getCategoriesInformation(Product $product, LanguageId $languageId): CategoriesInformation
```

Progi1984 commented on src/Core/Category/NameBuilder/CategoryDisplayNameBuilder.php:
```diff
@@ -0,0 +1,265 @@
+    /**
+     * @var CategoryRepository
+     */
+    private $categoryRepository;
```
```suggestion
    protected $categoryRepository;
```

Progi1984 commented on src/Core/Category/NameBuilder/CategoryDisplayNameBuilder.php:
```diff
@@ -0,0 +1,265 @@
+    /**
+     * @var string
+     */
+    private $breadcrumbSeparator;
```
```suggestion
    protected $breadcrumbSeparator;
```

Progi1984 commented on src/Core/Category/NameBuilder/CategoryDisplayNameBuilder.php:
```diff
@@ -0,0 +1,265 @@
+     *
+     * @return CategoryId[]
+     */
+    private function getDuplicateNameIds(ShopId $shopId, LanguageId $languageId, bool $useCache): array
```
```suggestion
    protected function getDuplicateNameIds(ShopId $shopId, LanguageId $languageId, bool $useCache): array
```

Progi1984 commented on src/Core/Category/NameBuilder/CategoryDisplayNameBuilder.php:
```diff
@@ -0,0 +1,265 @@
+     *
+     * @return array<int, string[]>
+     */
+    private function getDuplicateCategoriesBreadcrumbs(array $categoryIds, ShopId $shopId, LanguageId $languageId, bool $useCache): array
```
```suggestion
    protected function getDuplicateCategoriesBreadcrumbs(array $categoryIds, ShopId $shopId, LanguageId $languageId, bool $useCache): array
```

Progi1984 commented on src/Core/Category/NameBuilder/CategoryDisplayNameBuilder.php:
```diff
@@ -0,0 +1,265 @@
+     *
+     * @return array<int, string[]>
+     */
+    private function fetchBreadcrumbs(array $categoryIds, LanguageId $languageId): array
```
```suggestion
    protected function fetchBreadcrumbs(array $categoryIds, LanguageId $languageId): array
```

Progi1984 commented on src/Core/Category/NameBuilder/CategoryDisplayNameBuilder.php:
```diff
@@ -0,0 +1,265 @@
+     *
+     * @return string
+     */
+    private function buildBreadcrumb(CategoryId $categoryId, array $duplicateCategoriesBreadcrumbs): string
```
```suggestion
    protected function buildBreadcrumb(CategoryId $categoryId, array $duplicateCategoriesBreadcrumbs): string
```

Progi1984 commented on src/Core/Category/NameBuilder/CategoryDisplayNameBuilder.php:
```diff
@@ -0,0 +1,265 @@
+     *
+     * @return string
+     */
+    private function extractBreadcrumbFromParts(array $parts, int $step): string
```
```suggestion
    protected function extractBreadcrumbFromParts(array $parts, int $step): string
```

Progi1984 commented on src/Core/Category/NameBuilder/CategoryDisplayNameBuilder.php:
```diff
@@ -0,0 +1,265 @@
+     *
+     * @return string
+     */
+    private function buildCacheKeyForNameIds(ShopId $shopId, LanguageId $langId): string
```
```suggestion
    protected function buildCacheKeyForNameIds(ShopId $shopId, LanguageId $langId): string
```

Progi1984 commented on src/Core/Category/NameBuilder/CategoryDisplayNameBuilder.php:
```diff
@@ -0,0 +1,265 @@
+     *
+     * @return string
+     */
+    private function buildCacheKeyForBreadcrumbs(ShopId $shopId, LanguageId $langId): string
```
```suggestion
    protected function buildCacheKeyForBreadcrumbs(ShopId $shopId, LanguageId $langId): string
```

Progi1984 commented on src/Core/Domain/Category/QueryResult/CategoryForTree.php:
```diff
@@ -40,9 +40,14 @@ class CategoryForTree
+     * @var string
      */
-    private $localizedNames;
+    private $name;
```
```suggestion
    protected $name;
```

Progi1984 commented on src/Core/Domain/Category/QueryResult/CategoryForTree.php:
```diff
@@ -40,9 +40,14 @@ class CategoryForTree
+    /**
+     * @var string
+     */
+    private $displayName;
```
```suggestion
    protected $displayName;
```

Progi1984 commented on src/Core/Domain/Product/Query/GetProductForEditing.php:
```diff
@@ -45,17 +46,24 @@ class GetProductForEditing
+     * @var LanguageId some data from this class is only used for reading and is not expected to be edited, so it
+     *                 is provided only in the language of the user interface, which is defined by this parameter
+     */
+    private $displayLanguageId;
```
```suggestion
    protected $displayLanguageId;
```

Progi1984 reviewed (dismissed):
Some remaining feedbacks and it's an approval.

HanaRebaiQA reviewed (commented):
Hello @zuk3975 

Thanks for the fixed issues detected by @khouloudbelguith 
I checked this PR and i found this issue : 

The path is incorrect when change the category under another root category 

1. Go to BO > Catalog > Categories > Add a new category
2. Add this Category under Home root
3. Go to Catalog > Products V2 > Add a new product 
4. Click on `Add categories` and add this created category 
5. Check that the path is OK
6. Come back to Categories page 
7. Add a new root category 
8. Edit the created category under this new root category 
9. Come back to Products V2 page and edit the product
10. See error 1 > Check that the path of category previously created is under the new root in Categories list (tags)  
11. See error 2 > Check that the path of category previously created is under the new root in Default Category drop-down

Issue
![image](https://user-images.githubusercontent.com/92912932/170306792-01b516c4-ecde-4df5-b315-e663c65be813.png)

Categories are changed under root1 (in this example)
 
![image](https://user-images.githubusercontent.com/92912932/170308743-df48670f-186d-4c48-8b55-91a8b4ab44c0.png)

![image](https://user-images.githubusercontent.com/92912932/170308681-c0a25845-f5df-4b32-83c2-36b3b4441362.png)

https://user-images.githubusercontent.com/92912932/170308761-1575fc6f-0b6c-4d05-a241-b005a816ab80.mp4



Could you please check and feedback ?

Thank you!

zuk3975 commented:
just so I don't forget, I found out whats the problem - so after category update it seems that the nleft, nright are not updated (which should be done when parent is changed).
So its not related with the multishop and neither this particular PR, but rather category update. Im checking if it can be solved easily or needs dedicated PR for that fix :v:

jolelievre commented:
Ooohh, good to know thanks @zuk3975 

Maybe it's just a missing `Category::regenerateEntireNtree` missing somewhere in the category update that could fix this, if it's more complex we'll have to open a dedicated issue because I guess it could happen even without two root categories or without multishop.

Let us know if it's easy to fix, if not I think we can validate the PR.

HanaRebaiQA reviewed (commented):
Hello @MatShir 

Thanks for your feedback. Actually, the issue is not related to multistore. 
That's right that the `Add new root category` is used with multistore behavior. So here is the steps of the issue when multistore is disabled : 

**Issue 1** : The same mentioned above

1. Go to BO > Catalog > Categories > Add a new category (In my example : Categ1) 
2. Add this Category under Home root
3. Add new categories with the same name (In my example : C1 three times)
4. Go to Catalog > Products V2 > Add a new product
5. Click on Add categories and add these created categories 
6. Check that the path is OK (Home > Categ1 > C1)
7. Come back to Categories page
8. Add a new category under Home root (In my example : Categ2) 
9. Edit the created categories (C1) under this new category (Categ2 instead of Categ1)
10. Come back to Products V2 page and edit the product
11. See error 1 > Check that the path of categories previously created is under the old parent Category (tags list)
12. See error 2 > Check that the path of categories previously created is under the old parent in Default Category drop-down
13. See error 3 > Check that the path of categories previously created is under the old parent in the list of tags when click to `Add categories `

Here is a detailed screen-record with attached screenshots


https://user-images.githubusercontent.com/92912932/171615326-3af6267d-8144-47a2-9949-85e412cf9f12.mp4

![image](https://user-images.githubusercontent.com/92912932/171615624-fa5454d2-4b1a-42e4-8cf4-8826229d7c3b.png)

![image](https://user-images.githubusercontent.com/92912932/171615651-6d27a1d9-5fd8-4bfc-9001-50a6985594a5.png)


**Issue 2** : The dowpdown list is bad displayed (randomly) 

1. Add a new categories 
2. Click on apply
3. Click on `save and publish`
4. The dorpdown list is displayed at the top (see the same screen-record above)

OK :heavy_check_mark: 
![image](https://user-images.githubusercontent.com/92912932/171616006-0614edf0-5a51-439f-a8c6-cbafb097059d.png)

NOK :x: 
![image](https://user-images.githubusercontent.com/92912932/171616020-43acfced-8f45-494f-8f97-9d8c071a2b92.png)

Otherwise, i didn't found new issues when check it again.
I'm waiting for your feedback, if theses issues need to be fixed on this PR.

Thank you!

zuk3975 committed:
fix wrong cache default value

zuk3975 committed:
check if parent changed to initiate categoryNtree regeneration

jolelievre reviewed (dismissed):
Thanks @zuk3975

kpodemski commented:
I have two comments; sorry if I'm asking about something that was already asked, but it's quite a big discussion there 😅 

1. It was always possible to have the same category with different names in different stores. It does not work right now on the new product page. I'm always getting a name from the default store. I don't want to block this PR so let me know if this will be handled some other time.
2. Every duplicated category name adds two more queries. There is one edge-case scenario where I think it can be problematic: a shop selling auto parts. Very often, there are categories with the same name, for instance, year of production.  Simple idea: maybe if there are more than X duplicates, we should only show "Category name (#ID)"?

PS: It is out of the scope of this PR, but I've generated 3k categories, and we are loading them all at once after opening a popup. Are there any plans to make the tree dynamic so we can load subcategories dynamically?

If the #1 is not for now I'll leave it with ✅

khouloudbelguith reviewed (commented):
Hello @zuk3975,

**First issue**
1. Go to BO > Catalog > Categories page
2. Create two categories with the same name and associated with the same parent, for example, `test` 
![image](https://user-images.githubusercontent.com/16067358/182121023-3bcd332d-dd9d-4140-ae25-c99dcec7495c.png)
3. Go to BO > Catalog > Products page & Add a new product V2
4. Click on "Add categories" > "categories-modal" is displayed
5. In the Search input > Search the last category created, in our case `test` 
6. See error: in the dropdown list > the category id is not appended at the end
![image](https://user-images.githubusercontent.com/16067358/182122862-1f49b0e0-3dc1-410e-84f7-8227d974edcd.png)

It should be
![image](https://user-images.githubusercontent.com/16067358/182128759-39f3da2d-47bf-4887-9b4e-98b311db8b27.png)

In the Spec, I found this `The user can search by category and the parent’s categories. The search result list displays the breadcrumb of the category.`

Ping @MatShir what do you think? is it a bug or normal behavior?

**Second issue**

**_Spec_**
_If the product is not associated with any category except the ‘Home’ branch, then ‘Home’’s child categories are displayed, just the first level of the tree is displayed/opened._

=> In my case, All the tree is opened & displayed

![image](https://user-images.githubusercontent.com/16067358/182136445-52123710-a37c-472b-8b2c-40dd5f72e55c.png)

I attached a screen record

https://user-images.githubusercontent.com/16067358/182136754-87bb0b06-5eb8-462e-a956-1af81a195652.mp4


**Third issue**

**_Spec_**
_Below the drop-list is display the breadcrumb of the product. Updating the main category updates the breadcrumb displayed below the list and the breadcrumb of the front office._

Did I miss something? Nothing is missing 
![image](https://user-images.githubusercontent.com/16067358/182138582-4d6c50c6-1715-4494-8690-1cefc9066f37.png)

Following these steps, I have a difference between the "the main category" and the "the breadcrumb of the front office"

1. Create those categories

`Home > CategoryFirstLevel > CategorySecondLevel` and `Home > CategoryFirstLevel > CategorySecondLevel`
![image](https://user-images.githubusercontent.com/16067358/182139507-e9ff5191-b9a2-43cb-9be7-45a691a0f158.png)
2. Disable `CategoryFirstLevel`
3. Go to BO > Catalog > Products V2
4. Add a new product
5. Click on Add categories and select the categories with the same name `CategorySecondLevel`
6. Change the Default category to `CategorySecondLevel` 
![image](https://user-images.githubusercontent.com/16067358/182139903-99be8c78-e48c-4517-b710-19c51b7f6ecd.png)
7. Save and preview the product in the FO
8. Check the  "the breadcrumb of the front office" = `Home / CategorySecondLevel / Book1record `
![image](https://user-images.githubusercontent.com/16067358/182140138-b75c642d-77e6-4e8d-85e7-b9aca666aade.png)

But in the BO, I have this `Home > CategoryFirstLevel > CategorySecondLevel (#26)`

It is not the same as described in the spec.

Ping @MatShir what do you think?

**Fourth issue**

1. Go to BO > Product V2 & add a new product
2. Click Add categories
3. Please select any other category & change it to the main category
4. Go to BO > Catalog > Categories and delete the main category by selecting this option "If they have no other category, I want to associate them with the parent category." or this option "If they have no other category, I want to delete them as well."
![image](https://user-images.githubusercontent.com/16067358/182145013-dfc0d6ec-3937-48a9-a964-a5dd6a44f473.png)
5. Go to Back to Catalog > Products & Edit the last product created
6. See error: We can remove the main category
7. Remove the tag of the main category & click on save
8. See error 
![image](https://user-images.githubusercontent.com/16067358/182145535-f92363ea-7600-4130-9d7f-0b53b1f98cf4.png)

I attached a screen record

https://user-images.githubusercontent.com/16067358/182145746-79ba7ff3-77ca-47ba-a844-2408b3b3c9f0.mp4

Thanks!

zuk3975 commented:
Note for 1. issue (search dropdown not showing ids).
Dropdown breadcrumbs are filled in javascript side and its quite hard determining which one needs an id.
I wonder if we could instead keep it consistent between search and the tags and show the display name the same in both?
Meaning that search dropdown would also contain the breadcrumb ONLY if there are categories with different names (would act abosolutely the same as the tags). It would be very simple to do and we would get some performace as in javascript we wouldn't need to rebuild the breadcrumbs (so 1 loop less). Just considering this approach, because it seems quite logical. :thinking:

zuk3975 committed:
fix categories expanding/collapsing and potential nullpointer error

zuk3975 commented:
2 issue fixed.

zuk3975 commented:
Note for 3 issue. 
Now It seems to work the same as it was previously :thinking: (could you approve that?). If its the case then the issue is not in scope of the pr anyway.

zuk3975 committed:
do not allow removing last category as well as default

zuk3975 commented:
1. issue (search dropdown not showing ids) probably needs @MatShir attention - I wonder if we could instead keep it consistent between search and the tags and show the display name the same in both?
Meaning that search dropdown would also contain the breadcrumb ONLY if there are categories with different names (would act abosolutely the same as the tags). It would be very simple to do and we would get some performance as in javascript we wouldn't need to rebuild the breadcrumbs (so 1 loop less). Just considering this approach, because it seems quite logical and we would reuse the backend code this way and avoid implementing some complicated piece of code in frontend which will do similar stuff. But if its not acceptable, then I can continue and try to make it work as defined :man_shrugging: 
2. issue (expand/collapse tree) - Fixed heavy_check_mark
3. issue - now It seems to work the same as it was previously thinking (could you approve that?). If its the case then the issue is not in scope of the PR
4. issue (deleting main category) - fixed. Not allowing to delete last category or if its a main category :heavy_check_mark:

NeOMakinG commented on admin-dev/themes/new-theme/js/pages/product/components/categories/categories-manager.ts:
```diff
@@ -85,18 +85,29 @@ export default class CategoriesManager {
   }
 
+  private extractCategoryPreview(tag: HTMLElement): string {
+    const tagPreviewElement = tag.querySelector(ProductCategoryMap.categoryNamePreview) as HTMLElement;
```
👿

zuk3975 replied:
yeah... old PR. Fixing it :runner:

NeOMakinG commented on admin-dev/themes/new-theme/js/pages/product/components/categories/categories-manager.ts:
```diff
@@ -85,18 +85,29 @@ export default class CategoriesManager {
         categories.push({
           id: Number(idInput.value),
           name: this.extractCategoryName(tag as HTMLElement),
+          displayName: this.extractCategoryPreview(tag as HTMLElement),
```
👿

zuk3975 committed:
fix ts type errors

NeOMakinG reviewed (dismissed):
Front lgtm 🍬

MatShir commented:
Issue 1: Very good idea, it will be clearer and more consistent to keep the logic.

zuk3975 committed:
show displayName instead of breadcrumb when searching in typeahead

zuk3975 commented:
> Issue 1: Very good idea, it will be clearer and more consistent to keep the logic.

Glad to hear :partying_face: . Done :heavy_check_mark: 
![Screenshot from 2022-08-10 11-50-27](https://user-images.githubusercontent.com/31609858/183858950-7610d9be-2d5d-40bb-b926-ddd7e8ec46f4.png)

jolelievre reviewed (dismissed):
Thanks @zuk3975

zuk3975 commented:
I found additional issue so Im putting it on WIP again to avoid wasting QA time.
As in screenshot bellow, the "clothes" has same parent and are identical so they should also have an id.

![Screenshot from 2022-10-04 17-07-12](https://user-images.githubusercontent.com/31609858/193841238-cbb17a46-e66b-4631-a872-dabb94778dd0.png)

zuk3975 commented:
> I found additional issue so Im putting it on WIP again to avoid wasting QA time. As in screenshot bellow, the "clothes" has same parent and are identical so they should also have an id.
> 
> ![Screenshot from 2022-10-04 17-07-12](https://user-images.githubusercontent.com/31609858/193841238-cbb17a46-e66b-4631-a872-dabb94778dd0.png)

As discussed with @MatShir this shouldn't be a problem and its not worth the time to try adjusting it, so we leave it as is :heavy_check_mark:

zuk3975 committed:
fix duplicated method after rebase

khouloudbelguith reviewed (commented):
Hello @zuk3975,

**Issue1**

1. Go to BO > Catalog > Categories > Create a category "CategoryA" & "CategoryB" & "CategoryC"
2. Go to BO > Catalog > Product V2 > Add a new product & Enable it
3. In the categories section add the "categoryA" & "CategoryB" & "CategoryC" and select the "categoryA" as a default category 
![image](https://user-images.githubusercontent.com/16067358/198015913-feb3febc-6637-4e8d-8f0d-d5dc168d489d.png)
4. Go to BO > Catalog > Categories > Delete the category "CategoryA"
5. Go to FO > Search the product & click on it 
6. See error 
![image](https://user-images.githubusercontent.com/16067358/198016288-8facbadf-b885-45e6-adea-a29fbe6bfebc.png)
7. Go back & edit the last product created from BO
8. See error, you can remove the "CategoryB" & "CategoryC" from the Categories section 
![image](https://user-images.githubusercontent.com/16067358/198016560-9a87d939-021d-40ef-bb0f-536a5e7bf8af.png)
9. try to remove them => See error 
![image](https://user-images.githubusercontent.com/16067358/198016644-517c8d0e-651d-463c-b82f-3b17f989e4b5.png)
10. Click on "Add categories"
11. See error => No modal displayed & an error in the console 
![image](https://user-images.githubusercontent.com/16067358/198016955-07b28126-d57a-4833-9e45-b41799c9203b.png)

**Issue2** I don't know if it is a normal behavior or a bug

**_Case1_**
1. Go to BO > Catalog > Categories > Create a category "home" Or "HOME"
2. Go to BO > Catalog > Product V2 > Add/Edit any product
3. In the categories section > select the "Home" category & the "HOME" category
4. Check the tags: **Home (`#2`)** & **Home > HOME** 
![image](https://user-images.githubusercontent.com/16067358/198019233-25250b55-c030-4fd5-aaff-75a4f6466be6.png)

In this case, are the categories considered with the same name?
PS: they have different parents.
Is this case Ok?

**_Case2_**
1. Go to BO > Catalog > Categories > Create a category "ACCESSORIES"
2. Go to BO > Catalog > Product V2 > Add/Edit any product
3. In the categories section > select the "Accessories" category & the "ACCESSORIES" category
4. Check the tags: **Home > Accessories** & **Home > ACCESSORIES**

![image](https://user-images.githubusercontent.com/16067358/198020164-cc3c98ad-5cc8-42f0-85ae-32bf5d9d8374.png)

In this case, are the categories considered with the same name?
PS: they have the same parent.
Is this case Ok?

**Issue3**

1. Create a disabled category
2. Go to BO > Catalog > Products V2 > Add a product
3. Click on Add categories
4. In the Modal
5. The disabled category is displayed in italic font-style

![image](https://user-images.githubusercontent.com/16067358/198088734-14bccbcf-09bc-4c93-b1d6-c9af0355bcf2.png)

Isn't the wanted behavior? ping @MatShir what do you think?

**Issue4**

1. Go to BO > Catalog > Products V2 > Add a product
2. Click on Add categories
3. Click on save and publish
4. The dropdown list is displayed randomly

I attached a screen record

https://user-images.githubusercontent.com/16067358/198090821-1bef3767-54a6-4b83-83e0-a39b5e64162b.mp4


Thanks!

zuk3975 commented:
the **1st** issue is a bug (i'll try to fix it).
Note for the 4th issue - the list is ordered by category id (so when the id is not visible it may seem to be random, but its not)
All other seems to be the expected behavior.

MatShir commented:
> the **1st** issue is a bug (i'll try to fix it). Note for the 4th issue - the list is ordered by category id (so when the id is not visible it may seem to be random, but its not) All other seems to be the expected behavior.
+1

zuk3975 commented:
I found additional issue related to product list. If product is not assigned to Home category it will not be shown in the list. (because default category id filter is set to PS_HOME_CATEGORY id.

zuk3975 commented:
okay, to be fair the roots of 1st issue are kind of not related to this PR.
I managed to reprodue the same with old page.
I created 2 new categories (catD and catE in screenshot)
, added them to a new product named "hello", set one as default (catD) and removed it in BO categories page. Then product ends up with no default category and brings all kinds of issues.
![Screenshot from 2022-11-04 16-28-49](https://user-images.githubusercontent.com/31609858/200000758-634308ba-2ad1-4988-8a24-8feb15ab7bf3.png)
![Screenshot from 2022-11-04 16-29-26](https://user-images.githubusercontent.com/31609858/200000755-a26065f0-8347-4655-bd24-5823df7f0235.png)
![Screenshot from 2022-11-04 16-29-42](https://user-images.githubusercontent.com/31609858/200000747-b88993d2-49f6-404b-8303-dea4dff1cb59.png)

NeOMakinG reviewed (approved):
front part lgtm enough 🍬

khouloudbelguith reviewed (commented):
Hi @zuk3975,

I found this issue reported previously here: https://github.com/PrestaShop/PrestaShop/pull/28141#pullrequestreview-1156969943

**Issue1**

1. Go to BO > Catalog > Categories > Create a category "CategoryA" & "CategoryB" & "CategoryC"
2. Go to BO > Catalog > Product V2 > Add a new product
3. In the categories section add the "categoryA" & "CategoryB" & "CategoryC" and select the "categoryA" as a default category 
![image](https://user-images.githubusercontent.com/16067358/198015913-feb3febc-6637-4e8d-8f0d-d5dc168d489d.png)
4. Go to BO > Catalog > Categories > Delete the category "CategoryA"
7. Go back & edit the last product created from BO
8. See error, you can remove the "CategoryB" & "CategoryC" from the Categories section 
![image](https://user-images.githubusercontent.com/16067358/198016560-9a87d939-021d-40ef-bb0f-536a5e7bf8af.png)
9. try to remove them => See error 
![image](https://user-images.githubusercontent.com/16067358/198016644-517c8d0e-651d-463c-b82f-3b17f989e4b5.png)
10. Click on "Add categories"
11. See error => No modal displayed & an error in the console 
![image](https://user-images.githubusercontent.com/16067358/198016955-07b28126-d57a-4833-9e45-b41799c9203b.png)

I attached a screen record

https://user-images.githubusercontent.com/16067358/201166644-38b18f4f-eeae-4d3b-b63f-6445d4929a43.mp4


I tried with Product V1 => OK

https://user-images.githubusercontent.com/16067358/201162001-23975d3d-ffe5-44eb-b03c-26a7c4f599b9.mp4

**Issue2**

1. Create a new category "Cat"
2. Add new product V2 & add this category as a default category
3. Edit the category created to "Cat_updated"
4. Go back to the last product created & edit it
5. In the categories section > See error, in the "Associated categories:" displays the old one.
6. Click on "Add categories"
7. In the "Associated categories" displays the old one.

I attached a screen record


https://user-images.githubusercontent.com/16067358/201166586-6fbad0c5-fe69-471a-9429-00e70ef54783.mp4


**Issue3**
https://github.com/PrestaShop/prestashop-specs/blob/master/content/1.7/back-office/catalog/products/add-edit-v2.md#catSpec: egories
_Spec_: `If the product is already associated with categories, then the categories branches are opened until the categories associated are displayed.`

In my case, I created a product with two categories selected "Home" and "Clothes"
If I click on the "Add categories" button, the sub-categories "Men" and "Women" are displayed
![image](https://user-images.githubusercontent.com/16067358/201166342-57436342-964b-418f-9b58-5514613ab7d6.png)

I think, the expected behavior 
![image](https://user-images.githubusercontent.com/16067358/201166386-b0c29b06-cb59-45e8-8d12-54c9867968d6.png)

Ping @MatShir what do you think?

Thanks!

zuk3975 commented:
1 issue as I already mentioned is old. (in product page v1 is the same, just try to save the product and you'll see the error that its missing the "default' category selection. So after pr https://github.com/PrestaShop/PrestaShop/issues/30219 it shouldn't happen anymore.
2 issue I couldn't reproduce :thinking: 
3 issue I feel like we are going circles, I already addressed similar issue months ago, its just hard to find the conversation. But I believe that was the expected behavior (to have one layer of tree opened when parent is selected) as I see from my change and a comment here https://github.com/PrestaShop/PrestaShop/pull/28141/commits/2a6a7253daeb40210851545e5df1b0e05869dadd#diff-bd21d700916a33322ef542dbc3672cb39f2d70f7c7f305e378e56859910f2b15R250

khouloudbelguith reviewed (approved):
Hi @zuk3975,

Thanks for your feedback.
As discussed in slack, the second issue is reproduced when multistore is enabled.
Multishop wasn't yet handled for categories at all, so it may be out of scope.

I'm approving this PR.

Thanks!

jolelievre commented:
Thanks @zuk3975 and @khouloudbelguith
